types/logger/logger.go

```diff
@@ -42,3 +46,61 @@ func (w funcWriter) Write(p []byte) (int, error) {
 
 // Discard is a Logf that throws away the logs given to it.
 func Discard(string, ...interface{}) {}
+
+type limitData struct {
+	lim          *rate.Limiter
+	lastAccessed time.Time
+	msgBlocked   bool
+}
+
+// RateLimitedFn implements rate limiting by fstring on a given Logf.
+// Messages are allowed through at a maximum of r messages/second, in
+// bursts of up to b messages at a time.
+func RateLimitedFn(logf Logf, f float64, b int) Logf {
+	r := rate.Limit(f)
+	msgList := make(map[string]limitData)
+	lastPurge := time.Now()
+
+	rlLogf := func(s string, args ...interface{}) {
+		if rl, ok := msgList[s]; ok {
+
+			// Fields of structs contained in maps can't be modified; this is
+			// the workaround. See issue https://github.com/golang/go/issues/3117
+			temp := msgList[s]
+			temp.lastAccessed = time.Now()
+			msgList[s] = temp
+
+			if rl.lim.Allow() {
+				rl.msgBlocked = false
+				logf(s, args)
+			} else {
+				if !rl.msgBlocked {
+					temp = msgList[s]
+					temp.msgBlocked = true
+					msgList[s] = temp
+					logf("Repeated messages were suppressed by rate limiting. Original message: " +
+						fmt.Sprintf(s, args))
+				}
+			}
+		} else {
+			msgList[s] = limitData{rate.NewLimiter(r, b), time.Now(), false}
+			msgList[s].lim.Allow()
+			logf(s, args)
+		}
+
+		// Purge msgList of outdated keys to reduce overhead. Must be done by copying
+		// over to a new map, since deleting in maps is done through a zombie flag
+		if time.Since(lastPurge) >= time.Minute {
+			newList := make(map[string]limitData)
+			for k, v := range msgList {
+				if time.Since(v.lastAccessed) < 5*time.Second {
```

Comment:
This algorithm can work, but it leaves the struct at an unpredictable size: if more messages get printed per minute, we'll use more memory. It would be nice to have our memory usage be as stable as possible. You could look at this module for an LRU implementation instead: github.com/golang/groupcache/lru. 

An LRU has the disadvantage that items might be evicted from the cache if a lot of distinct format strings are used. On the other hand, that isn't the sort of bug we're defending against; if there are 100 messages that all need rate limiting, we're doomed anyway. An LRU would keep the most common format strings at the top of the list (least likely to get evicted), and the least common format strings at the bottom (most likely to be evicted, but doesn't matter because they don't need rate limiting because they're least common).

Reply:
I definitely like the idea of the LRU! I looked into the provided library though and it also just calls `delete` on its map (that holds the cache) when removing the oldest elements, which doesn't solve the memory problem. The implementation seems fairly simple though (and it gets even simpler if we use the idea for the rate limiter) so I'll put together a rate limiter version that actually purges the memory we want released.